docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,314 @@
+* Fetching data directly from the API using `apiclient`
+* Rendering config files (or just strings) using `schnauzer`
+
+For open-source tools, Bottlerocket prefers the rendered config file approach, as it doesn't require patching upstream code in order to consume the Bottlerocket API.
+First-party tools have no such constraints, and so there exists tools that take either approach.
+
+Because settings are discovered at runtime on the system, and indeed may no longer be handled by models written in Rust, it will no longer be feasible to create an `apiclient` with the same strong typing as exists today.
+It will instead be preferred that system services, including first-party services, all route their fetching of API settings through their own configuration files, which will be rendered by Bottlerocket whenever the related settings are changed.
+
+Today, configuration templates are rendered by `schnauzer` using [handlebars](https://handlebarsjs.com/) templates.
+`schnauzer` fetches the entirety of the settings tree, defines some useful helper functions, and then uses these as context for rendering the template.
+With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
+These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `---` as such:
+
+```toml
+---
+[required-extensions]
+hostname = v1
+kubernetes = v1
+---
+# Below here the template is written as usual
+```
+
+This will allow Bottlerocket to find the owning extensions for the data needed to render the template.
+It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "usptream" changes.
+
+`schnauzer` can also continue to be invoked as a setting generator without the need for a templated configuration file.
+In those circumstances, you are still required to list your extension dependencies and their version, but these can be done as command line arguments:
+
+```bash
+$ schnauzer --required-extension 'hostname=v1' '{{ network.hostname }}'
+```
+
+As an optimization, templates which render to `systemd` units will invoke a restart command to systemd
+
+#### Extending Handlebars Helpers
+Handlebars has a few utilities for applying simple processes to data in order to format it appropriately for your document; however, you often need to implement your own "helpers" for more complex text formatting.
+This is currently done in Bottlerocket by adding helper implementations to the `schnauzer` tool directly.
+Bottlerocket settings extensions will expose versioned handlebar helpers:
+
+```bash
+# hostname.toml setting extension config file
+[templating.v1] # template helpers are versioned
+helpers = [
+    "localhost_aliases"
+]
```

Comment:
Something seems off here. It looks like there would be multiple versions of the schema by which you can specify which helpers exist.

Maybe this? I guess I'm not super clear on the versioning of helpers.

```suggestion
[templating.helpers] # template helpers are versioned
localhost_aliases = "v1"
```

Reply:
Each existing version of a setting would have its own helpers that it provides. During template rendering, we request a settings at a given version, we need to ensure that the template helpers at that version continue to exist and provide the same API.